upcoming/composable-pipelines-pattern.md

```diff
@@ -0,0 +1,392 @@
+
+The Generator functionality in python basically means that you create a function that, rather than returning a single data structure once (say for example a list of items), rather will return a generator object which can later be iterated over by repeadetly calling its next method, or by the shorthand `for item in ...` syntax.
+
+What is special about a generator compared to other *iterables* in python such as lists (which also support the `for item in ...` syntax), is that the generator function will start evaluating itself and yielding its output objects one by one only after iteration has started. Thus, minimal intermediate data is created and stored in memory.
+
+To give an example, say that we have a file, chr_y.fa containing a little bit of the familiar A, C, G, T DNA nucleotides from the human Y chromosome, in the ubiquotous [FASTA file format](http://en.wikipedia.org/wiki/FASTA_format):
+
+**chr_y.fa:**
+````fasta
+>Homo_sapiens.GRCh37.67.dna_rm.chromosome.Y
+GTGATTGTTGTTAGGTTCTTAGCTGCTTCTGAAAAATGGGGTGATAATCTTAGAAGGACT
+TGCTTCATGGGATGTGGTCCATAAAACTTCCTCTGCCCCAGTTGTAGGGCAGAAGACAAT
+TTCTGTTACTGTAGTTTGGCCTTTTTTTGCAGAGATTCAGACATCTGTTTACTGACCTTA
+GTTAAATTGTGACACTATGCCTAAAGGAGCCTGCAAGCTTTTATTTTTGCTCACTATGAA
+GTCATCATTCAATTGTAAAATTTCCTTTTTAAGTTTCAGGTTGACTTAATGTCTGTCAAA
+GCACAGTCTTTGGCAATAACAAAACAAACATATGCTGAATGAAAATGTTTAAGAGATGGA
+TGACTATTTACTACTAAAAGAAGAAAAATTGGAAGAGAATAAAATGAAAACATGCATCTC
+CTAAACCATATGNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNTGCATGTTTTAG
+TGATTCATACTAGGTCAGTATTATAAAACTATGCTTTGTCCTTGTAAGGGGAGGCTTAAA
+````
+*(The N:s mean that the nucleotides at those positions are not known, and the first line, starting with '>', is just a label, that should be skipped in our case)*
+
+*(For the real world Human Y-Chromosome fasta file, see [this link](http://ftp.ensembl.org/pub/release-67/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.67.dna_rm.chromosome.Y.fa.gz) [68MB, gzipped])*
+
+Then we can read the content of the file line by line and process it in sequential steps, using chained generators:
+
+````python
+# Create a lazy-evaluated file reader, just yielding the
+# lines of the file one by one, as we later iterate on
+# the generation function that it returns.
+def fasta_reader(filename):
+	with open(filename,'r') as infile:
+		for line in infile:
+			yield line
+
+# A base complementer, which converts each nucleotide in a
+# DNA sequence into its "complement" nucleotide (the one it
+# will sit paired to in the double-strand DNA helix molecule)
+def base_complementer(line_generator):
+	for line in line_generator:
+		if line[0] != '>':
+			translation_table = {
+			  'A' : 'T',
+			  'T' : 'A',
+			  'C' : 'G',
+			  'G' : 'C',
+			  'N' : 'N',
+			  '\n' : '\n'}
+			for nuc in line:
+				newline.append(translation_table[nuc])
+			yield newline
```

Comment:
Yeah, I think I figured out where this weirdness came from: I was using the atom editor, because of it's markdown preview, but it turns out to be super-buggy when pasting formatted code (but what to expect from an editor written in Node.js and not Go, after all :) ). I had to modify it manually afterwards, which was not a good thing. Edited with a more stable editor now.